features/futures/pci-passthrough.md

```diff
@@ -0,0 +1,85 @@
+---
+title: PCI passthrough support
+layout: default
+design_doc: true
+revision: 1
+status: proposed
+---
+
+Introduction
+------------
+
+GPU passthrough is already available in XAPI, this document proposes to also
+offer passthrough for all PCI devices through XAPI.
+
+Design proposal
+---------------
+
+New methods for PCI object:
+- `PCI.enable_dom0_access`
+- `PCI.disable_dom0_access`
+- `PCI.get_dom0_access_status`: compares the outputs of `/opt/xensource/libexec/xen-cmdline`
+  and `/proc/cmdline` to produce one of the four values that can be currently contained
+  in the `PGPU.dom0_access` field:
+  - disabled
+  - disabled_on_reboot
+  - enabled
+  - enabled_on_reboot
+
+  How do determine the expected dom0 access state:
+  If the device id is present in both `pciback.hide` of `/proc/cmdline` and `xen-cmdline`: `enabled`
+  If the device id is present not in both `pciback.hide` of `/proc/cmdline` and `xen-cmdline`: `disabled`
+  If the device id is present in the `pciback.hide` of `/proc/cmdline` but not in the one of `xen-cmdline`: `disabled_on_reboot`
+  If the device id is not present in the `pciback.hide` of `/proc/cmdline` but is in the one of `xen-cmdline`: `enabled_on_reboot`
+
+  A function rather than a field makes the data always accurate and even accounts for
+  changes made by users outside XAPI, directly through `/opt/xensource/libexec/xen-cmdline`
+
+If wanted:
+The first two methods would only be callable if the `enable-pci-passthrough` setting is set
+to a yes/true value in XAPI's configuration files. Default: no.
+
+With these generic methods available, comes the question of the existing methods and fields
+on the PGPU object:
+- `PGPU.enable_dom0_access`
+- `PGPU.disable_dom0_access`
+- `PGPU.dom0_access` (DB field)
+
+They can all be deprecated, or retained. If retained, the contents of `PGPU.dom0_access`
```

Comment:
> They can all be deprecated, or retained

Not sure what this means, deprecating means they still exist and need to be functional for old clients relying on them. The only way to not retain them is to remove them.

The problematic element is the database field. Anything relying on fetching the PGPU record and reading the field from it will need it to be accurate.

To maintain this the PCI function needs to be able to set the database field whenever is called on a PGPU, and the PGPU functions need to be equivalent to that. Some module to dump the shared functionality might be present.

I do not recommend adding PGPU.get_dom0_access_status, the PCI function should be enough

Reply:
Yes the idea is to deprecate the PGPU methods and DB field while **still** keeping them up to date for the existing use.